students/kegan/session10/MailroomTools.py

```diff
@@ -0,0 +1,489 @@
+
+    def thank(self, all_donations=False):
+        """ Returns personalized thank you message for this donor.
+        Thanks donor for all donations if all_donations is True.
+        Args:
+            all_donations (bool) : thank donor for all donations
+        Returns:
+            str : thank you message for donor
+        """
+        num = self.num if all_donations else 1
+        first = self.donations[:-1] if all_donations else []
+        rest = self.donations[-1]
+        substrings = {
+            'donor': self.name,
+            's': 's' if num > 1 else '',
+            'and': ' and ' if num > 1 else '',
+            'first': ', '.join(['${:,.2f}'.format(d) for d in first]),
+            'rest': '${:,.2f}'.format(rest),
+            'totalling':
+                '' if num < 2 else
+                ', totalling {}{},'.format(
+                    'an incredible ' if self.total > 500
+                    else '', '${:,.2f}'.format(self.total))}
+        message = \
+            'Dear {donor},\nThank you for your generous gift{s} of ' +\
+            '{first}{and}{rest}. Your donation{s}{totalling} will ' +\
+            'go towards feeding homeless kittens in Seattle. ' +\
+            'From the bottom of our hearts, we at Miuvenile Care thank you.' +\
+            '\n\nRegards,\nBungelina Bigglesnorf\nChairwoman, Miuvenile Care'
+        return message.format(**substrings)
+
+    def add(self, *donations):
+        """ Adds passed donations to this donor.
+        Args:
+            donations (args) : donations as arguments
+        """
+        donations = self.intake_donations(*donations)
+        self._donations.extend(donations)
+
+    def __str__(self):
+        """ Returns this donor as a string.
+        Returns:
+            str : donor as string
+        """
+        donations = ', '.join([
+            '${:,.2f}'.format(d) for d in self.donations])
+        return '{}: {}'.format(self.name, donations)
+
+    def __repr__(self):
+        """ Returns representation of this donor.
+        Returns:
+            str : representation of this donor
+        """
+        donations = ', '.join([
+            str(d) for d in self.donations])
+        return 'Donor("{}", {})'.format(self.name, donations)
+
+    def __eq__(self, other):
+        """ Returns whether this donor is equal to other in
+        both name and all donations.
+        Args:
+            other (Donor) : donor to compare
+        Returns:
+            bool : True if other donor is equal to this donor
+        """
+        return (
+            self.name == other.name and
+            self.donations == other.donations)
+
+    def __lt__(self, other):
+        """ Returns whether this donor is less than
+        other. Compares name and donations.
+        Args:
+            other (Donor) : donor to compare
+        Returns:
+            bool : True if this donor is less than other
+        """
+        return (
+            self.name < other.name and
```

Comment:
careful here.

by using "and", one donor has to have a name less than AND fewer donations. the name less than is good -- that's alphabetical order.

But also requiring the donations to be less means you can have odd results if one name is greater than the other, but the donations are less. I prototyped this, add this is what I get:

```
In [33]:     d1 = Donor("Abe", [2])
    ...:     d2 = Donor("Bob", [1])
    ...: 

In [34]: d2 < d1
Out[34]: False

In [35]: d1 < d2
Out[35]: False
```

What you probably want is to check the name, and only if the names are equal, then test the donations. you could do this with a bit of logic, but the easiest way is probably to put them in a tuple:

return (self.name, self.donations) < (other.name, other.donations)

tuples are sorted in order of the items -- only if the first item is the same does the second one get checked.

Also, probably sorting by total donations would make more sense, as if you sort on teh list, the firsdt one dominates:

```
In [36]: [1,2,3] < [10,20,30]
Out[36]: True
```
all good.

```
In [37]: [1,2000] < [10,20,30]
Out[37]: True
```
maybe not what you'd want

Finally, of the only reason to implement comparisons is to get a good sort order, implementing a sort_key function that can be used will result in faster sorting:

```
    @staticmethod
    def sort_key(self):
        return (self.name, self.donations)
```

Then: 

sorted(list_of_donors, key=Donor.sort_key)